internal/registry/registry.go

```diff
@@ -0,0 +1,241 @@
+type repoPullConfig struct {
+	repo        string
+	authOptions []crane.Option
+}
+
+//go:generate mockgen -source=registry.go -package=registry -destination=mock_registry_api.go
+
+type Registry interface {
+	ImageExists(ctx context.Context, image string, po ootov1alpha1.PullOptions) (bool, error)
+	ExtractToolkitRelease(v1.Layer) (*DriverToolkitEntry, error)
+	GetLayersDigests(ctx context.Context, image string) ([]string, *repoPullConfig, error)
+	GetLayerByDigest(digest string, pullConfig *repoPullConfig) (v1.Layer, error)
+}
+
+type registry struct{}
+
+func NewRegistry() Registry {
+	return &registry{}
+}
+
+func (r *registry) ImageExists(ctx context.Context, image string, po ootov1alpha1.PullOptions) (bool, error) {
+	_, _, err := r.getImageManifest(ctx, image, &po)
+	if err != nil {
+		te := &transport.Error{}
+		if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
+			return false, nil
+		}
+		return false, fmt.Errorf("could not get image %s: %w", image, err)
+	}
+	return true, nil
+}
+
+func (r *registry) GetLayersDigests(ctx context.Context, image string) ([]string, *repoPullConfig, error) {
+	manifest, pullConfig, err := r.getImageManifest(ctx, image, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get manifest from image %s: %w", image, err)
+	}
+
+	digests, err := r.getLayersDigestsFromManifestStream(manifest)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get layers digests from manifest of the image %s: %w", image, err)
+	}
+
+	return digests, pullConfig, nil
+}
+
+func (r *registry) GetLayerByDigest(digest string, pullConfig *repoPullConfig) (v1.Layer, error) {
+	return crane.PullLayer(pullConfig.repo+"@"+digest, pullConfig.authOptions...)
+}
+
+func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, error) {
+	var found bool
+	dtk := &DriverToolkitEntry{}
+	obj, err := r.getHeaderFromLayer(layer, driverToolkitJSONFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find file %s in image layer: %w", driverToolkitJSONFile, err)
+	}
+
+	dtk.KernelFullVersion, found, err = unstructured.NestedString(obj.Object, "KERNEL_VERSION")
+	if !found || err != nil {
+		return nil, fmt.Errorf("failed to get KERNEL_VERSION from %s, found %t: %w", driverToolkitJSONFile, found, err)
+	}
+
+	dtk.RTKernelFullVersion, found, err = unstructured.NestedString(obj.Object, "RT_KERNEL_VERSION")
+	if !found || err != nil {
+		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from %s, found %t: %w", driverToolkitJSONFile, found, err)
+	}
+
+	dtk.OSVersion, found, err = unstructured.NestedString(obj.Object, "RHEL_VERSION")
+	if !found || err != nil {
+		return nil, fmt.Errorf("failed to get RHEL_VERSION from %s, found %t: %w", driverToolkitJSONFile, found, err)
+	}
+	return dtk, nil
+}
+
+func (r *registry) getImageManifest(ctx context.Context, image string, po *ootov1alpha1.PullOptions) ([]byte, *repoPullConfig, error) {
```

Comment:
I was already confused in v1 why this method returns the pull configuration i.e. the crane options.
Could we extract that code into a separate function that only generates the options?

Reply:
fixed